packages/service-client/odsp-client/src/odspClient.ts

```diff
@@ -0,0 +1,205 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { IDocumentServiceFactory } from "@fluidframework/driver-definitions";
+import {
+	OdspDocumentServiceFactory,
+	OdspDriverUrlResolver,
+	createOdspCreateContainerRequest,
+} from "@fluidframework/odsp-driver";
+import {
+	type ContainerSchema,
+	DOProviderContainerRuntimeFactory,
+	IFluidContainer,
+	FluidContainer,
+	IRootDataObject,
+} from "@fluidframework/fluid-static";
+import {
+	AttachState,
+	IContainer,
+	IFluidModuleWithDetails,
+} from "@fluidframework/container-definitions";
+import { IClient } from "@fluidframework/protocol-definitions";
+import { Loader } from "@fluidframework/container-loader";
+import { v4 as uuid } from "uuid";
+import {
+	IOdspResolvedUrl,
+	OdspResourceTokenFetchOptions,
+} from "@fluidframework/odsp-driver-definitions";
+import { ITokenResponse } from "@fluidframework/azure-client";
+// eslint-disable-next-line import/no-deprecated
+import { requestFluidObject } from "@fluidframework/runtime-utils";
+import { IRequest } from "@fluidframework/core-interfaces";
+import {
+	OdspClientProps,
+	OdspConnectionConfig,
+	OdspContainerServices,
+	OdspServiceAttributes,
+} from "./interfaces";
+import { OdspAudience } from "./odspAudience";
+
+/**
+ * OdspClient provides the ability to have a Fluid object backed by the ODSP service within the context of Microsoft 365 (M365) tenants.
+ *
+ * @alpha @sealed
+ */
+export class OdspClient {
+	private readonly documentServiceFactory: IDocumentServiceFactory;
+	private readonly urlResolver: OdspDriverUrlResolver;
+
+	public constructor(private readonly properties: OdspClientProps) {
+		const getSharePointToken = async (options: OdspResourceTokenFetchOptions) => {
+			const tokenResponse: ITokenResponse =
+				await this.properties.connection.tokenProvider.fetchStorageToken(
+					options.siteUrl,
+					"",
+				);
+			return {
+				token: tokenResponse.jwt,
+			};
+		};
+
+		const getPushServiceToken = async (options: OdspResourceTokenFetchOptions) => {
+			const tokenResponse: ITokenResponse =
+				await this.properties.connection.tokenProvider.fetchOrdererToken(options.siteUrl);
+			return {
+				token: tokenResponse.jwt,
+			};
+		};
+		this.documentServiceFactory = new OdspDocumentServiceFactory(
```

Comment:
I'd love to see mistakes of AzureClient being addressed. We need a way for the client to provide their own factory / build adapters on top of it.
The moment it's out, we will tell people they can use caching capabilities built into the driver, but to do so - they need to pass policy and cache implementation to driver factory. How users will be able to do so without cloning this code?

We need to follow principle - simple things should be easy, complex things should be possible.
Or - there have to be escape path.

Reply:
TODO: adding this as follow-up